include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1289 @@
+{
+	return irr_glsl_blinn_phong_cos_eval_DG_wo_clamps(NdotH, NdotV*NdotV, NdotL*NdotL, n, a2);
+}
+float irr_glsl_blinn_phong_cos_eval_DG(in irr_glsl_LightSample s, in irr_glsl_IsotropicMicrofacetCache microfacet, in irr_glsl_IsotropicViewSurfaceInteraction i, in float n, in float a2)
+{
+	return irr_glsl_blinn_phong_cos_eval_DG(microfacet.NdotH, i.NdotV, s.NdotL, n, a2);
+}
+
+float irr_glsl_blinn_phong_cos_eval_DG(in float NdotH, in float TdotH, in float BdotH, in float NdotL, in float TdotL, in float BdotL, in float NdotV, in float TdotV, in float BdotV, in float nx, in float ny, in float ax2, in float ay2)
+{
+	return irr_glsl_blinn_phong_cos_eval_DG_wo_clamps(NdotH, NdotH * NdotH, TdotH * TdotH, BdotH * BdotH, TdotL * TdotL, BdotL * BdotL, TdotV * TdotV, BdotV * BdotV, NdotV * NdotV, NdotL * NdotL, nx, ny, ax2, ay2);
+}
+float irr_glsl_blinn_phong_cos_eval_DG(in irr_glsl_LightSample s, in irr_glsl_AnisotropicMicrofacetCache microfacet, in irr_glsl_AnisotropicViewSurfaceInteraction i, in float nx, in float ny, in float ax2, in float ay2)
+{
+	return irr_glsl_blinn_phong_cos_eval_DG(
+		microfacet.isotropic.NdotH,
+		microfacet.TdotH,
+		microfacet.BdotH,
+		s.NdotL,
+		s.TdotL,
+		s.BdotL,
+		i.isotropic.NdotV,
+		i.TdotV,
+		i.BdotV,
+		nx, ny, ax2, ay2
+	);
+}
+
+#ifdef GEN_CHOICE_STREAM
+void instr_eval_and_pdf_execute(in instr_t instr, in MC_precomputed_t precomp, inout irr_glsl_LightSample s, inout irr_glsl_AnisotropicMicrofacetCache _microfacet)
+{
+	uint op = instr_getOpcode(instr);
+
+	//speculative execution
+	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+	params_t params = instr_getParameters(instr, bsdf_data);
+	uint ndf = instr_getNDF(instr);
+	float a = params_getAlpha(params);
+	float a2 = a*a;
+#ifndef ALL_ISOTROPIC_BXDFS
+	float ay = params_getAlphaV(params);
+	float ay2 = ay*ay;
+#endif
+	float bxdf_eval_scalar_part = 0.0;
+	float pdf = 0.0;
+
+	const mat2x3 ior = bsdf_data_decodeIoR(bsdf_data, op);
+	const mat2x3 ior2 = matrixCompMult(ior, ior);
+	const bool is_bsdf = !op_isBRDF(op); //note it actually tells if op is BSDF or BUMPMAP or SET_GEOM_NORMAL (divergence reasons)
+
+#ifndef NO_TWOSIDED
+	handleTwosided(instr, s, _microfacet);
+#endif
+
+	float cosFactor = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
+
+	irr_glsl_AnisotropicMicrofacetCache microfacet;
+	bool is_valid = true;
+#ifndef NO_BSDF
+	//here actually using stronger check for BSDF because it's probably worth it
+	if (op_isBSDF(op) && irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
+	{
+		float orientedEta, rcpOrientedEta;
+		irr_glsl_getOrientedEtas(orientedEta, rcpOrientedEta, currInteraction.isotropic.NdotV, ior[0].x);
+		is_valid = irr_glsl_calcAnisotropicMicrofacetCache(microfacet, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, orientedEta, rcpOrientedEta);
+	}
+	else
+#endif
+	{
+		microfacet = _microfacet;
```

Comment:
actually you could change the name of `is_valid` to `can_eval` and here you could do
`can_eval = cosFactor>FLT_MIN;`